Port THS to ATen.

ezyang opened the pull request:
The basic structure of the patch:

- All kernels in aten/src/THS got rewritten as native
  functions in aten/src/ATen/native/sparse

  I took the liberty to rename some of the kernels,
  opting for a longer, more transparent names than
  things like 'spaddcmul'.

- Instead of holding fields for sparse tensor in the TH
  C struct THSTensor, they are now held in a C++ class
  SparseTensorImpl (this explains why I had to do this
  all in one go; I can't have *two* reps for sparse
  tensors!)

  Along the way, we change a key internal representation
  invariant: an "empty" sparse tensor has dimI == 1 and
  dimV == 0 (this is different from dimI == 0 and dimV == 0
  we had before); this ensures that we maintain the invariant
  that dim == dimI + dimV.  "Scalar" sparse tensors are
  made illegal, because there really is no way to properly
  express them in COO format.

- Because we haven't ported THCS or any of the traditional
  dense TH implementations, there is a new set of adapter
  functions in native/LegacyBridge.cpp exclusively devoted
  to deciding whether or not to go to the new native implementation
  or back to the legacy TH binding (prefixed with th_).
  The intent is that when everything gets ported, we can
  delete this file.

- I've kept the stubs for all the THS functions, but they now all
  error if you try to actually call them.  Eventually, we should
  replace these with calls to ATen so that everything keeps
  working.

- I gobbled up SparseMM (SparseMM.cpp is no more). It was tasty.

There are some miscellaneous improvements which were needed for other
changes in this patch:

- There is now AT_FORALL_SCALAR_TYPES_EXCEPT_HALF, which does what
  it says on the tin.

- axpy templated function moved to BlasUtils.h, there's a new macro
  which lets you easily forward to all of the TH functions. We also expose
  THBlas_(copy).  I'm not terribly pleased with these functions but
  they seem to serve a purpose they need.

- New method on Tensor to get TensorImpl*, unsafeGetTensorImpl

- accessor() is now this-const, since const-correctness on Tensor is a lie

- New toSparse()/toDense() methods on Type; now you can call these
  directly without having to manually apply at::toSparse/toDense
  on the Backend and then running toBackend yourself.

Changes to the kernels:

- Previously, the whole body of all kernels was compiled for
  every supported scalar type.  In our new implementation,
  the scalar dispatch has been pushed into the smallest extent
  which (1) is not in a type loop and (2) requires statically
  knowing the scalar type.  These sites all use
  AT_DISPATCH_ALL_TYPES.  I tried to use lambdas as much as
  possible, but sometimes it was not possible when a OpenMP
  pragma was used.

- Anywhere we tested if the nDimension of a tensor was zero,
  we replaced with a test that numel is zero.  Because, as we
  known, nDimension of zero-size tensors in TH is zero, and
  that's wrong wrong wrong (and not done this way in ATen).

Some subtleties:

- Places where previously fastget1d was used, I now use a
  TensorAccessor.  However, you have to be careful about grabbing
  the accessor, because sometimes you will be accessor'ing
  indices/values and they are empty, which means they will
  be *1D* ("oh, aren't indices always 2D?" Nope. Nyet.)
  So, essentially, it is only safe to grab an accessor *after*
  you have checked that nnz != 0.  All of these shenanigans
  will go away when we properly support zero-size dimensions.

  A few places, we test for this case just by wrapping the loop
  in a conditional on nnz.  Some other places this is not so easy,
  so we instead short-circuit the function with a special case for
  when nnz == 0 (usually, these implementations are degenerate).

- There is a very subtle but important difference between
  _sparse_get_impl(self)->indices() and self._indices();
  the latter may return a view!  This is because nnz is
  not guaranteed to match the dimensions of indices/values;
  you can "truncate" a sparse tensor by setting the nnz.
  Actually, I think this is not a good idea and we should
  enforce a stronger invariant, but for this patch I slavishly
  adhere to the old ways, and as such I have to be very
  careful if I want to resize something, I had better use
  the former and not the latter.

- I had to reimplement broadcasting by hand (thus the s_
  and non-s_ functions in the sparse native files).  There
  is a very important distinction between foo_out and foo_,
  so it is important that the LegacyBridge function always
  call to the lower layer, and not try to avoid boilerplate
  by calling to another LegacyBridge function first.
  I did NOT put broadcasting in LegacyBridge (even though,
  ultimately, that's where it must live), because the th_
  functions which are invoked from LegacyBridge handle
  broadcasting themselves, and I don't want to broadcast
  twice.

- Sparse function MUST explicitly specify the Type they
  dispatch from, otherwise Variable wrapping/unwrapping will
  not work correctly.  If you use _get_sparse_impl, that is
  sufficient to levy this requirement.

- The "has native" tests in LegacyBridge.cpp are not 100%,
  because some of the functions are mixed dense-sparse functions,
  and so you can't just say, "Oh, if it's sparse and CPU, call
  the native sparse implementation."  This is handled on a
  case by case basis.  There is some especially complex
  logic for add(), which has dense-dense, sparse-sparse
  and dense-sparse implementations.

- I added some uses of SparseTensorRef in native_functions.yaml,
  but you will notice that these are all on native_* functions,
  and not the actual, top-level functions.  So the SparseTensorRef
  is purely documentary (helping you not call the wrong overload)
  but there is no magic; we do the wrapping ourselves the hard
  way. (This is in constrast to the TH binding code which is magical.)
  Except for _sparse_mask; _sparse_mask is magical.

- There is a raw_copy_sparse_ method, which is really my way of
  getting around the fact that copy_ has never been implemented
  for sparse tensors (even before this patch), but there IS a
  super secret, internal way of doing these copies that the THS
  code used, and which I needed to get my hands on when I did this
  port.  We should refactor so that either (a) copy_ does support
  sparse-sparse copy natively, or (b) we do this other ways.

- Irritatingly, I must explicitly resize_as_ before copy_ into
  a tensor.  This was not the case with THTensor_(copy) but I don't
  have any direct binding that doesn't have this requirement.

- For some reason, the sparse tensor constructor accepts a scalar
  tensor for the values tensor.  This is kind of weird because
  you always need an nnz-dimension.  However, the old code supported
  this and just expanded it into a 1D size 0 tensor; so we need some
  explicit code to do this.

There are maybe a bit more AT_ASSERTs in some of the kernels
than is wise.  I added them all when I was debugging and was
loathe to remove them.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

TODO: I claim that I have "error"ed out all of the old THS functions, but this has not actually been done yet except for one function.

ezyang committed:
Port THS to ATen.

The basic structure of the patch:

- All kernels in aten/src/THS got rewritten as native
  functions in aten/src/ATen/native/sparse

  I took the liberty to rename some of the kernels,
  opting for a longer, more transparent names than
  things like 'spaddcmul'.

- Instead of holding fields for sparse tensor in the TH
  C struct THSTensor, they are now held in a C++ class
  SparseTensorImpl (this explains why I had to do this
  all in one go; I can't have *two* reps for sparse
  tensors!)

  Along the way, we change a key internal representation
  invariant: an "empty" sparse tensor has dimI == 1 and
  dimV == 0 (this is different from dimI == 0 and dimV == 0
  we had before); this ensures that we maintain the invariant
  that dim == dimI + dimV.  "Scalar" sparse tensors are
  made illegal, because there really is no way to properly
  express them in COO format.

- Because we haven't ported THCS or any of the traditional
  dense TH implementations, there is a new set of adapter
  functions in native/LegacyBridge.cpp exclusively devoted
  to deciding whether or not to go to the new native implementation
  or back to the legacy TH binding (prefixed with th_).
  The intent is that when everything gets ported, we can
  delete this file.

- I've kept the stubs for all the THS functions, but they now all
  error if you try to actually call them.  Eventually, we should
  replace these with calls to ATen so that everything keeps
  working.

- I gobbled up SparseMM (SparseMM.cpp is no more). It was tasty.

There are some miscellaneous improvements which were needed for other
changes in this patch:

- There is now AT_FORALL_SCALAR_TYPES_EXCEPT_HALF, which does what
  it says on the tin.

- axpy templated function moved to BlasUtils.h, there's a new macro
  which lets you easily forward to all of the TH functions. We also expose
  THBlas_(copy).  I'm not terribly pleased with these functions but
  they seem to serve a purpose they need.

- New method on Tensor to get TensorImpl*, unsafeGetTensorImpl

- accessor() is now this-const, since const-correctness on Tensor is a lie

- New toSparse()/toDense() methods on Type; now you can call these
  directly without having to manually apply at::toSparse/toDense
  on the Backend and then running toBackend yourself.

Changes to the kernels:

- Previously, the whole body of all kernels was compiled for
  every supported scalar type.  In our new implementation,
  the scalar dispatch has been pushed into the smallest extent
  which (1) is not in a type loop and (2) requires statically
  knowing the scalar type.  These sites all use
  AT_DISPATCH_ALL_TYPES.  I tried to use lambdas as much as
  possible, but sometimes it was not possible when a OpenMP
  pragma was used.

- Anywhere we tested if the nDimension of a tensor was zero,
  we replaced with a test that numel is zero.  Because, as we
  known, nDimension of zero-size tensors in TH is zero, and
  that's wrong wrong wrong (and not done this way in ATen).

Some subtleties:

- Places where previously fastget1d was used, I now use a
  TensorAccessor.  However, you have to be careful about grabbing
  the accessor, because sometimes you will be accessor'ing
  indices/values and they are empty, which means they will
  be *1D* ("oh, aren't indices always 2D?" Nope. Nyet.)
  So, essentially, it is only safe to grab an accessor *after*
  you have checked that nnz != 0.  All of these shenanigans
  will go away when we properly support zero-size dimensions.

  A few places, we test for this case just by wrapping the loop
  in a conditional on nnz.  Some other places this is not so easy,
  so we instead short-circuit the function with a special case for
  when nnz == 0 (usually, these implementations are degenerate).

- There is a very subtle but important difference between
  _sparse_get_impl(self)->indices() and self._indices();
  the latter may return a view!  This is because nnz is
  not guaranteed to match the dimensions of indices/values;
  you can "truncate" a sparse tensor by setting the nnz.
  Actually, I think this is not a good idea and we should
  enforce a stronger invariant, but for this patch I slavishly
  adhere to the old ways, and as such I have to be very
  careful if I want to resize something, I had better use
  the former and not the latter.

- I had to reimplement broadcasting by hand (thus the s_
  and non-s_ functions in the sparse native files).  There
  is a very important distinction between foo_out and foo_,
  so it is important that the LegacyBridge function always
  call to the lower layer, and not try to avoid boilerplate
  by calling to another LegacyBridge function first.
  I did NOT put broadcasting in LegacyBridge (even though,
  ultimately, that's where it must live), because the th_
  functions which are invoked from LegacyBridge handle
  broadcasting themselves, and I don't want to broadcast
  twice.

- Sparse function MUST explicitly specify the Type they
  dispatch from, otherwise Variable wrapping/unwrapping will
  not work correctly.  If you use _get_sparse_impl, that is
  sufficient to levy this requirement.

- The "has native" tests in LegacyBridge.cpp are not 100%,
  because some of the functions are mixed dense-sparse functions,
  and so you can't just say, "Oh, if it's sparse and CPU, call
  the native sparse implementation."  This is handled on a
  case by case basis.  There is some especially complex
  logic for add(), which has dense-dense, sparse-sparse
  and dense-sparse implementations.

- I added some uses of SparseTensorRef in native_functions.yaml,
  but you will notice that these are all on native_* functions,
  and not the actual, top-level functions.  So the SparseTensorRef
  is purely documentary (helping you not call the wrong overload)
  but there is no magic; we do the wrapping ourselves the hard
  way. (This is in constrast to the TH binding code which is magical.)
  Except for _sparse_mask; _sparse_mask is magical.

- There is a raw_copy_sparse_ method, which is really my way of
  getting around the fact that copy_ has never been implemented
  for sparse tensors (even before this patch), but there IS a
  super secret, internal way of doing these copies that the THS
  code used, and which I needed to get my hands on when I did this
  port.  We should refactor so that either (a) copy_ does support
  sparse-sparse copy natively, or (b) we do this other ways.

- Irritatingly, I must explicitly resize_as_ before copy_ into
  a tensor.  This was not the case with THTensor_(copy) but I don't
  have any direct binding that doesn't have this requirement.

- For some reason, the sparse tensor constructor accepts a scalar
  tensor for the values tensor.  This is kind of weird because
  you always need an nnz-dimension.  However, the old code supported
  this and just expanded it into a 1D size 0 tensor; so we need some
  explicit code to do this.

There are maybe a bit more AT_ASSERTs in some of the kernels
than is wise.  I added them all when I was debugging and was
loathe to remove them.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Move expand outside of dispatch so autograd works.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
I know why the broadcast tests are failing (the broadcast calls are behind a dispatch: SparseCPU, so they aren't interpreted as part pieces I need to autograd through) but there is some special casing for this situation in TH which I can't seem to quite replicate when some overloads are native while others are TH. I'll try to work it out tomorrow.

ezyang committed:
Make the addmm overload actually work by not using SparseTensorRef

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Hack to duplicate the derivatives for our now two definitions. Mercifully the derivatives are short.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Apparently, TH has a special case to make foo_ functions method only, and if you don't do this the Python arg parsing is wrong. Ick.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
dynamic_type is Scalar instead of real, hack around it

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Unfuck the tracing situation

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Apply DCE to a test_jit case, fixes wobbling

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update test_function

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
There are some very uncomfortable hacks in the last set of commits. Please review with care.

ezyang committed:
Start killing the old THSTensor functions.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Some last mile fixes for dispatch confusion in sparse_coo_tensor functions.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1399,15 +1399,23 @@
+# the variants, you must call this on type directly.
+#
+# Maybe this meant we were supposed to take a dtype as an argument here.
+# Hmmmmm.
```
That's not unreasonable, although it wouldn't be enough for this case, because neither argument is the actual type you want the dispatch to happen on in the end.

Actually, when C10 dispatcher lands none of this will matter because you won't have to pick an argument to dispatch on.

gchanan replied:
Also I think just taking a Type or ScalarType would work (as you suggest) -- you can do all the same logic that is implemented in tensor_new here.

ezyang replied:
I attempted this patch, but I triggered a codegen bug which I don't feel like fixing:

```
In file included from torch/csrc/autograd/generated/python_torch_functions.cpp:23:0:
torch/csrc/autograd/generated/python_torch_functions_dispatch.h: In function ‘at::Tensor torch::autograd::dispatch__native_sparse_co
o_tensor_unsafe(const at::Tensor&, const at::Tensor&, at::IntList)’:
torch/csrc/autograd/generated/python_torch_functions_dispatch.h:584:25: error: ‘dtype’ was not declared in this scope
   maybe_initialize_cuda(dtype);
```

ezyang replied:
https://github.com/ezyang/pytorch-1/tree/snapshot/taking-type-for-sparse-coo-tensor

ezyang committed:
Simplified regression test for ONNX failures

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Increase tolerance on super resolution test

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Error out _move

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Some gotchas about Python argument parsing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
More robust dynamic_type normalization, fixes ONNX bug

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
More stubs

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
lintfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
I give up on giving BC hints

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
lintfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Kill the rest of the dead stuff.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
bugfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gchanan commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -0,0 +1,64 @@
+  }
+  indices_ = indices;
+  values_ = values;
+  nnz_ = empty ? 0 : values.size(0);
```
will values.size(0) not be 0 in the empty case?

ezyang replied:
Yes, but it's an accident. I think it's clearer this way. I added this comment:

```
  // TODO: Eliminate this ternary when we handle size zero dimensions.
  // (Actually, this will "accidentally" work today because all zero-size
  // tensors have size [0], and so you'll get 0 when empty is zero; but it's
  // more explicit this way.)
```

gchanan commented on aten/src/ATen/SparseTensorImpl.h:
```diff
@@ -1,10 +1,83 @@
+  std::vector<int64_t> size_;
+  // INVARIANT: indices_.size(1) >= nnz_ && values_.size(0) >= nnz_
+  int64_t nnz_ = 0;
+  int64_t dimI_ = 0; // number of sparse dimensions
```
should we rename these?

ezyang replied:
Maybe. Note that this name is exposed semi-publically as `_dimI()` method. Do you have good suggestions? `_sparseDims` and `_denseDims`?

gchanan replied:
we can still keep them around for legacy reasons.  I like the names you suggested.

gchanan commented on aten/src/ATen/SparseTensorImpl.h:
```diff
@@ -1,10 +1,83 @@
+
+  // A sparse tensor is 'coalesced' if every index occurs at most once in
+  // the indices tensor, and the indices are in sorted order.
+  // Most math operations can only be performed on ordered sparse tensors
```
what does ordered mean here?  do you just mean coalesced?

ezyang replied:
yep, coalesced

gchanan commented on aten/src/ATen/SparseTensorImpl.h:
```diff
@@ -1,10 +1,83 @@
+  // WARNING: This function does NOT preserve invariants of dimI/dimV with
+  // respect to indices and values
+  void raw_resize_(int64_t dimI, int64_t dimV, ArrayRef<int64_t> size) {
+    // UGHHHHH.  Legacy special case
```
you could future proof this code with USE_TH_ZERO_SIZE_DIM as in https://github.com/pytorch/pytorch/pull/8468 if you want.

ezyang replied:
sure, if you get the tests to pass before I merge this XD

ezyang commented on aten/src/ATen/ScalarType.h:
```diff
@@ -8,6 +8,27 @@
+// to be.  (Probably, the order /shouldn't/ matter, but I couldn't pin down
+// where someone had hardcoded a number.)
+
+#define AT_FORALL_SCALAR_TYPES_EXCEPT_HALF(_) \
```
Sure I'll move it to BlasUtils.h

ezyang commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -0,0 +1,64 @@
+namespace at {
+
+// SparseTensorImpl defaults to a [0] size tensor with one sparse dimension and
+// no dense dimensions.  This is kind of arbitrary but we want the math to work
```
```
// SparseTensorImpl defaults to a [0] size tensor with one sparse dimension and
// no dense dimensions.  This is kind of arbitrary but we want an empty tensor
// to have one dimension (zero dimensions means its scalar), and we have
// the invariant that dim == dimI + dimV, so *one* of the inner dimensions
// has to be one.
```

ezyang committed:
Code review.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Expunge is_variable_or_undefined

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gchanan commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -0,0 +1,70 @@
+
+namespace at {
+
+// SparseTensorImpl defaults to a [0] size tensor with one sparse dimension and
```
I find this comment a little confusing; can you structure it as:
- This is what it _should_ be (i.e.size [1,0])
- Since we don't have support for that, this is what we chose to do instead

ezyang replied:
```
// An empty dense tensor defaults to a 1-dimensional tensor of size [0]
// (recall, it is not a 0-dimensional tensor, because such a tensor would
// a scalar and have one element)
//
// Thus, an empty sparse tensor should be a 1-dimensional tensor of size [0].
// Furthermore, we have dim == sparseDims + denseDims; since this is a sparse
// tensor, let us say that an empty sparse tensor has sparseDims == 1 and
// denseDims == 0.  (There is a degree of freedom here, but given that this
// is a sparse dimension, it seems reasonable to demand that sparseDims > 0).
//
// In an ideal world, this would then mean we allocate a [1,0] size indices
// tensor and a [0] size values tensor for such an empty tensor.  However,
// we don't currently support zero-size dimensions, so we can't actually
// do this; so we just allocate zero-size tensors for everything.
```

Actually, there is a bit in the comment defending why we should default sparseDims == 1, denseDims == 0, I reworded it for clarity.

gchanan commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -0,0 +1,70 @@
+  return dimI_ + dimV_;
+}
+Scalar SparseTensorImpl::localScalar() {
+  AT_ERROR("sparse tensors cannot be scalars");
```
I don't think this is right; the requirement in dense is that numel() == 1, i.e. this is basically a conversion.  We should do the same thing here.

ezyang replied:
It appears to be broken in master (I see localScalar is used in `is_nonzero`):

```
>>> import torch
>>> i = torch.LongTensor([[0, 2], [1, 0], [1, 2]])
>>> v = torch.FloatTensor([3,      4,      5    ])
>>> y = torch.sparse.FloatTensor(i.t(), v)
>>> y.is_nonzero()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: numel is not implemented for type torch.sparse.FloatTensor
```

but seems reasonable enough to fix.

ezyang replied:
Fixed a lot of it; while working on it I realized that "scalar" sparse tensors actually make sense as a coherent concept, but I didn't go far enough to fix code to handle this correctly.

gchanan commented on aten/src/ATen/SparseTensorImpl.h:
```diff
@@ -1,10 +1,84 @@
+
+  // WARNING: This function does NOT preserve invariants of dimI/dimV with
+  // respect to indices and values
+  void raw_resize_(int64_t dimI, int64_t dimV, ArrayRef<int64_t> size) {
```
maybe, as in https://github.com/pytorch/pytorch/pull/8468, we should add `legacy` to the name?

ezyang replied:
agreed

ezyang replied:
Oh fuck it, I'm going to make scalar sparse tensors work

gchanan commented on aten/src/ATen/copy_wrapper.py:
```diff
@@ -96,12 +96,12 @@
+# message which claims something is not supported, when it actually is.  But
+# the correct fix in this case was to NOT call copy_from
 FUNCTION_FROM_SWAP = CodeTemplate("""\
 Tensor & ${Type}::_s_copy_from(const Tensor & src, Tensor & dst, bool non_blocking) const {
```
I'm confused -- how is this function useful now?

ezyang replied:
There's some hidden code that you can't see here which overrides `_s_copy_from` for CUDA tensors. So the general strategy is `_s_copy_from` is called from `s_copy` if a cpu-cuda conversion is involved; it's the double dispatch function.

gchanan commented on aten/src/ATen/native/BlasUtils.h:
```diff
@@ -0,0 +1,45 @@
+#include <TH/THBlas.h>
+#include <ATen/ScalarType.h>
+
+// This header file shouldn't be anything permanent; it's just a temporary
```
Could just move this specialization into non-generic TH (hpp). I did something similar with the Copy functions in THC, that way you don't have to invent new ATen APIs.  But this looks fine to me if you want to keep it.

ezyang replied:
Moving into THC sounds fine.

gchanan commented on aten/src/ATen/native/LegacyBridge.cpp:
```diff
@@ -0,0 +1,469 @@
+  // an all or nothing affair, because the internal representation
+  // is different
+  static bool _has_native(const Tensor& self) {
+    return self.is_sparse() && !self.is_cuda();
```
nit: _type_has_native(self.type())

gchanan commented on aten/src/ATen/native/LegacyBridge.cpp:
```diff
@@ -0,0 +1,469 @@
+// In principle, there are some degrees of freedom in how we could gotten to our
+// native CPU sparse implementations.  In particular, the handling of how we are
+// simulating multiple dispatch is asymmetric with how th_add_out handles the
+// multiple dispatch.  We justify this in two ways: (1) knowing when to
```
could you explain how it's asymmetric?

ezyang replied:
```
// Note [Multiple dispatch to sparse]
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// In an ideal world, we would use direct support for multiple dispatch to
// say that add(Dense, Dense) should dispatch to one function, while
// add(Dense, Sparse) should dispatch to another function.
//
// In a world where we only have single dispatch, we can single dispatch on
// the first function, and then do an is_sparse() test on the second argument
// to direct ourselves to the correct argument.
//
// We are in neither of those worlds.  Instead, we have a th_add function
// which has legacy implementations in the single dispatch world, BUT our
// actual add function needs to call s_native_add if the function *would have*
// utilized a kernel implemented in THS (CPU sparse tensors).
//
// th_add is "good old single dispatch" which internally handles the is_sparse()
// test and also handles broadcasting.  s_native_add works asymmetrically:
// it doesn't handle broadcasting at all, and it ASSUMES that the relevant
// argument is a CPU sparse tensor.  Why the asymmetry?  It turns out it is not
// so easy to figure out if a kernel is implemented in THS; it's not as simple
// as testing if the first argument is CPU and sparse, because, e.g.,
// in add(Dense, Sparse), the sparse kernel is in the second argument.  So,
// the trampoline function is going to know about the overloads *anyway*; it
// might as well also handle is_sparse() and broadcasting while it's at it.
//
// Why not change TH to follow this new scheme?  We could... but since it's
// all going away when we finish porting the TH functions to ATen, we haven't
```

New comment

gchanan commented on aten/src/ATen/native/LegacyBridge.cpp:
```diff
@@ -0,0 +1,469 @@
+    auto other_sparse = other.is_sparse();
+    if (self_sparse && other_sparse) {
+      Tensor b_self, b_other;
+      std::tie(b_self, b_other) = expand_outplace(self, other, "add_out");
```
this actually works for sparse :)?  I'm pleasantly surprised.

ezyang replied:
I bet it only actually works if it's a dense tensor that gets broadcasted, cuz I'm pretty sure we don't have expand on sparse :) (in any case the tests didn't see anything wrong).

gchanan commented on aten/src/ATen/native/LegacyBridge.cpp:
```diff
@@ -0,0 +1,469 @@
+      // confusing, and we should have given these overloads different names.
+      // For now, we do it this way for consistency with the TH bindings
+      // (not that it is terribly consistent anyway).
+      return native_add_out(result, self, SparseTensorRef(other), alpha);
```
I'll admit I don't understand why this calls `native_add_out` but the above calls `s_native_add_out` -- the SparseTensorRef stuff makes sense though.

ezyang replied:
It's pretty simple, actually; in the old Declarations.cwrap, there was no broadcasting enabled for dense-sparse addition, so I mindlessly copied the behavior over.

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -73,11 +74,15 @@ signature.
 - `std::array<bool,N>` (where N is `1-4`).  NB: you MUST NOT put a space after the comma, otherwise
   this argument will not parse correctly.  (If you decide to fix this, make sure you fix the
   argument parser both in ATen and in PyTorch.)
+- `*` is a special sentinel argument, which doesn't translate into a real
```
can you not used the word `real` -- that has another meaning an a similar context :).  Maybe something like "an actual", "a realized" or similar.

gchanan commented on aten/src/ATen/native/README.md:
```diff
@@ -232,6 +237,44 @@ is that if you know that you will only ever be called with `Tensor`, a
+
+Tensor& s_add_(Tensor& self, const Tensor& other) {
+  // non-broadcasting implementation of inplace addition
+}
```
nice!

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1164,3 +1164,383 @@
+# Hmmmmm.
+
+- func: native_sparse_coo_tensor(IndexTensor indices, Tensor values) -> Tensor
+  variants: []
```
wait -- what does this mean?

ezyang replied:
`variants: []` means, don't generate `at::native_sparse_coo_tensor` or `Tensor:: native_sparse_coo_tensor`, but do create a method on `Type`. But I think I'm going to see if I can just make these take a `Type`.

gchanan replied:
I think we should consider methods on types to be an implementation detail, it doesn't match up with the (non-legacy) pytorch API.

gchanan commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -0,0 +1,449 @@
+
+namespace {
+  bool _is_same_size_as_sparse(const SparseTensor& self, const SparseTensor& src) {
+    return self._dimI() == src._dimI() && self._dimV() == src._dimV() && self.sizes().equals(src.sizes());
```
I think I called this `is_congruent` somewhere.

gchanan commented on aten/src/THS/generic/THSTensor.cpp:
```diff
@@ -65,218 +51,81 @@ THTensor *THSTensor_(newValues)(const THSTensor *self) {
-  THSTensor *self = (THSTensor *)THAlloc(sizeof(THSTensor));
-  THSTensor_(rawInit)(self);
-  return self;
+  THError("Internal error! THSTensor_(new)() shouldn't be called; use dtype.tensor() instead");
```
why did we keep these?  I agree we could eventually have these call ATen -- but what's the point of not just failing at compile time?

ezyang replied:
I kept it so that, if we eventually have these call ATen, all of the templates will still be here so it's an easier patch to do. Otherwise whoever does it has to dig out the old files, wire them all back up, and then start doing replacements.

gchanan commented on test/test_jit.py:
```diff
@@ -358,6 +358,16 @@ def fn(x, y):
+        def fn(x, y):
+            return x - y
+        trace, _ = torch.jit.get_trace_graph(fn, (x, y), nderivs=0)
+        print(trace)
```
do you want to print the trace in the test?  It's fine, just curious.

ezyang replied:
Nope, it's an accident

gchanan commented on tools/autograd/gen_python_functions.py:
```diff
@@ -613,17 +614,43 @@ def group_declarations(declarations):
+    # at the source but I have never understood what dynamic_type is
+    # supposed to be.
+    def normalized_dynamic_type(arg):
+        if arg['dynamic_type'] == 'real':
```
why do you actually need this?  I don't really know what `dynamic_type` is trying to get across here, so it's not obvious to me that it's wrong.  Can you just change is_coord_smaller to do what you want?

ezyang replied:
I can't only change `is_coord_smaller`, because there are equality tests done on `dynamic_type`, and if `real` and `Scalar` don't compare as equal I get the wrong sort. Maybe I should just fix this at the source and make sure everything works out OK. What do you think?

gchanan replied:
not sure what you mean by at the source, do you mean so the native and non-native declarations entries are the same?  I find it hard to argue about this without really understanding what `dynamic_type` is supposed to represent.

ezyang replied:
Yep, that's what I mean. Pinging @zdevito about the meaning of `dynamic_type` versus `type`.

gchanan commented on tools/autograd/gen_variable_type.py:
```diff
@@ -44,6 +44,18 @@
 }
 
+# These functions have their names recorded under trace renamed,
+RENAME_TRACE = {
```
is there a plan to get rid of these?

ezyang replied:
Yeah, the plan on record is https://github.com/pytorch/pytorch/issues/8495

gchanan commented on tools/autograd/gen_variable_type.py:
```diff
@@ -159,6 +171,19 @@ def should_trace(declaration):
     base_name = name[:-1] if declaration['inplace'] else name[:-4] if name.endswith('_out') else name
     if base_name in DONT_RECORD_TRACE:
         return False
+    # We need to disable these because their inner implementations implement
```
what's the plan moving forward with this?  How difficult is it to do something like 'we know this is a native function calling into a non-native function that does broadcasting...'.

ezyang replied:
Also #8495

ezyang committed:
Make new_with_tensor_sparse more CUDA safe

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Note about CUDA-safety in SparseTensorImpl

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Rename dimI/dimV to sparseDims/denseDims.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Clarify SparseTensorImpl comment.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make localScalar on SparseTensorImpl work.

Some other fixes too:

- Make numel uniformly supported on all types, not just dense
  types
- Add tests for is_nonzero() method (which exercises localScalar)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
CR

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Work around torch._thnn import problem

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Move BlasUtils into TH

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
lintfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Clarify Multiple dispatch to sparse

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gchanan reviewed (approved):
I don't think I fully understand what this will look like once everything is ported over, but this looks good to merge if you want.

ezyang committed:
CR

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Remove unnecessary print

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Disable constant tests.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Revert "Work around torch._thnn import problem"

This reverts commit 69398c6ee9e232f7004bfe821b155f2e61d1b83b.
